service/apis/qsl_service.go

```diff
@@ -0,0 +1,418 @@
+	// r := regexp.MustCompile(`([a-zA-Z]+)\[(\@[\,\@\"\=a-zA-Z0-9\-\.]*)\]\{([\*|[\,\@\"\=a-zA-Z0-9\-]*)`)
+	// r := regexp.MustCompile(`([a-zA-Z]+)\[(?:(\@[\,\@\"\=a-zA-Z0-9\-\.\|\:_]*|\*))\]\{([\*|[\,\@\"\=a-zA-Z0-9\-]*)`)
+	r := regexp.MustCompile(block_regex)
+	matches := r.FindStringSubmatch(splitquery[0])
+	log.Debugf("matches %#v\n", matches)
+	log.Debugf("%#v\n", r.SubexpNames())
+
+	if len(matches) < 2 {
+		log.Error("Malformed Query received: " + query)
+		return "", errors.New("Malformed Query: " + query)
+	}
+
+	// extract the values of the form objtype[filters]fields and assign to individual variables
+	objtype := strings.Title(matches[1])
+	filters := matches[2]
+	fields := matches[3]
+	_ = fields
+
+	log.Debugf("helperobjtype %#v\n", objtype)
+
+	log.Debugf("objecttype %#v\n", objtype)
+	// get a list of the fields supoorted for this object type
+	metafieldslist, err := qa.metaSvc.GetMetadataFields(objtype)
+	if err != nil {
+		log.Error("err in getting metadata fields")
+		log.Error(err)
+		return "", errors.New("Failed to connect to dgraph to get metadata")
+	}
+
+	if len(metafieldslist) == 0 {
+		log.Error("metadata for " + objtype + " not found in db. Will not be able to use * or find relationships")
+	}
+
+	log.Debugf("metafields for %s: %#v\n", objtype, metafieldslist)
+
+	// convert the filters and fields to corresponding dgraph lines for the query
+	fd, ff, err := CreateFiltersQuery(filters)
+	if err != nil {
+		return "", err
+	}
+	fl, err := CreateFieldsQuery(fields, metafieldslist, 0)
+	if err != nil {
+		return "", err
+	}
+
+	log.Debugf("Objtype: %#v\n", objtype)
+	log.Debugf("Filterfunc: %#v\nFilterdec: %#v\n", ff, fd)
+	log.Debugf("Fieldlist: %#v\n", fl)
+
+	// replace the filters and object type and add the list of fields
+	basequery[0] = strings.Replace(basequery[0], "$FILTERSDEC", fd, -1)
+	basequery[1] = strings.Replace(basequery[1], "$FILTERSFUNC", ff, -1)
+	basequery[1] = strings.Replace(basequery[1], "$OBJTYPE", objtype, -1)
+	basequery = append(basequery[0:2], fl...)
+
+	// if there's relations, this length will be greater than 1
+	if len(splitquery) > 1 && !(string(fields[0]) == "*" && len(fields) > 1) {
+		// recursively create the intermediate relations
+		middlefilter, err := qa.CreateDgraphQueryHelper(splitquery[1:], 1, objtype)
+		if err != nil {
+			return "", err
+		}
+		// and add them to the end of this list with the closing braces
+		basequery = append(basequery, middlefilter...)
+		basequery = append(basequery, []string{"}", "}"}...)
+	} else {
+		// just add closing braces
+		basequery = append(basequery, []string{"}", "}"}...)
+	}
+
+	// create the string joining all the lines with newlines
+	finalquery := strings.Join(basequery, "\n")
+	// log.Infof("New Query: %s\n", finalquery)
+	return finalquery, nil
+
+}
+
+// Run query on a dgraph instance
+func (qa *QSLService) ExecuteDgraphQuery(query string) (map[string]interface{}, error) {
```

Comment:
Please move ExecuteDgraphQuery() to dbclient.go, we put all DB operations to that file so no need to have "host" field on your QSLService struct.